src/realm/sync/noinst/client_reset_recovery.cpp

```diff
@@ -0,0 +1,933 @@
+        {
+            REALM_ASSERT(index != uint32_t(-1));
+            size_t lst_size = list.size();
+            uint32_t translated_from, translated_to;
+            bool allowed_to_move =
+                m_recovery_applier->m_lists.at(m_list_path)
+                    .move(static_cast<uint32_t>(index), m_instr.ndx_2, lst_size, translated_from, translated_to);
+            if (allowed_to_move) {
+                m_instr.prior_size = static_cast<uint32_t>(lst_size);
+                m_instr.path.back() = translated_from;
+                m_instr.ndx_2 = translated_to;
+            }
+            else {
+                do_not_resolve();
+            }
+        }
+
+    private:
+        Instruction::ArrayMove& m_instr;
+    };
+    Instruction::ArrayMove instr_copy = instr;
+    if (ArrayMoveResolver(this, instr_copy).resolve() == RecoveryResolver::Status::Success) {
+        InstructionApplier::operator()(instr_copy);
+    }
+}
+
+void RecoverLocalChangesetsHandler::operator()(const Instruction::ArrayErase& instr)
+{
+    struct ArrayEraseResolver : public RecoveryResolver {
+        ArrayEraseResolver(RecoverLocalChangesetsHandler* applier, Instruction::ArrayErase& instr)
+            : RecoveryResolver(applier, instr, "ArrayErase")
+            , m_instr(instr)
+        {
+        }
+        void on_list_index(LstBase& list, uint32_t index) override
+        {
+            auto obj = list.get_obj();
+            uint32_t translated_index;
+            bool allowed_to_delete =
+                m_recovery_applier->m_lists.at(m_list_path).remove(static_cast<uint32_t>(index), translated_index);
+            if (allowed_to_delete) {
+                m_instr.prior_size = static_cast<uint32_t>(list.size());
+                m_instr.path.back() = translated_index;
+            }
+            else {
+                do_not_resolve();
+            }
+        }
+
+    private:
+        Instruction::ArrayErase& m_instr;
+    };
+    Instruction::ArrayErase instr_copy = instr;
+    if (ArrayEraseResolver(this, instr_copy).resolve() == RecoveryResolver::Status::Success) {
+        InstructionApplier::operator()(instr_copy);
+    }
+}
+
+void RecoverLocalChangesetsHandler::operator()(const Instruction::SetInsert& instr)
+{
+    struct SetInsertResolver : public RecoveryResolver {
+        SetInsertResolver(RecoverLocalChangesetsHandler* applier, Instruction::SetInsert& instr,
+                          const std::string_view& instr_name)
+            : RecoveryResolver(applier, instr, instr_name)
+        {
+        }
+        void on_set(SetBase&) {}
+    };
+    static constexpr std::string_view instr_name("SetInsert");
+    if (!check_links_exist(instr.value)) {
+        m_logger.warn("Discarding a %1 which links to a deleted object", instr_name);
+        return;
+    }
+    Instruction::SetInsert instr_copy = instr;
+    if (SetInsertResolver(this, instr_copy, instr_name).resolve() == RecoveryResolver::Status::Success) {
```

Comment:
What is it exactly we get out of running the `RecoveryResolver::resolve()` first?

Reply:
We discussed this offline, but for posterity, the essence of the answer is that the RecoveryResolver has to track which list indices are accessed when the operation applies to a property of some embedded object across a list path. Then the indices can be transformed to apply to the new state, or the list can be tracked for verbatim copy.